File: Kakao/2019인턴/badUser.js
let answer = 0;
let visited = [];
let size;
let user_id;
let banned_id;
let map = new Map();

function solution(_user_id, _banned_id) {
    user_id = _user_id;
    banned_id = _banned_id;
    //console.log(banned_id);
    size = banned_id.length;

    matching(0);
    

    

    return answer;
}

function check(banUser, user){

    if(banUser.length !== user.length) return false;

    for(let i =0;i<banUser.length;i++){
        if(banUser[i] === '*')
            continue;
        
        if(banUser[i] !== user[i])
            return false;
    }
    return true;

}

function matching(depth){
    if(depth === size){
        //console.log(visited);
        let key = '';
        for(let i = 0;i<visited.length;i++){
            if(visited[i])
                key+= user_id[i];
        }
        //console.log(key);
        if(!map[key]){
            map[key] = 1;
            answer++;
        }
        
        return;
    }

    let banUser = banned_id[depth];
    //console.log();
    
    for(let i = 0; i<user_id.length;i++){
        if(visited[i]) continue;
        //console.log(depth,banUser,user_id[i]);
        if(check(banUser,user_id[i])){
            
            visited[i] = true;
            matching(depth+1);
            visited[i] = false;
        }
    }

}


(function solve(){
    let user_id = ["frodo", "fradi", "crodo", "abc123", "frodoc"];
    let banned_id = ["*rodo", "*rodo", "******"];
    
    console.log(solution(user_id,banned_id));

})();

File: Kakao/2019인턴/badUser.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

vector<vector<int> > bans;
bool check[8] = {false};
int answer = 0;
int userSize;
unordered_map<string,bool> map;

bool isSame(string ban,string user){
    if(ban.size() != user.size()) return false;
    
    for(int i =0;i<ban.size();i++){
        char bc = ban.at(i);
        char uc = user.at(i);
        if(bc == '*') continue;
        if(bc != uc) return false;
    }
    //<<ban<<endl;
    return true;

}


void selects(int depth){
    if(depth == bans.size()){
        
        string num = "";
        for(int i =userSize-1;i>=0;i--){
            if(check[i]){
                string ss = to_string(i);
                num += ss;
            }
        }
        //cout<< num<<endl;
        if(map.find(num) == map.end()){
            answer++;
            map.insert({num,true});
        }
        return;
    }
    
    vector<int> temp = bans[depth];

    for(int i =0;i<temp.size();i++){
        int index = temp[i];
        if(!check[index]){
            check[index] = true;
            selects(depth+1);
            check[index] = false;
        }
    }

}



int solution(vector<string> user_id, vector<string> banned_id) {
    userSize = user_id.size();

    for(int i =0;i<banned_id.size();i++){
        string banid = banned_id[i];
        //cout<<banid<<endl;
        vector<int> tempban;
        for(int j =0;j<user_id.size();j++){
            string userid = user_id[j];
            
            if(isSame(banid,userid)){
                tempban.push_back(j);
            }
        }
        bans.push_back(tempban);
        tempban.clear();
    }
    
    selects(0);


    return answer;
}